### PR TITLE
OADP-73: Usage

### PR DESCRIPTION
**Pull request by apinnick**
https://issues.redhat.com/browse/OADP-73

4.6+

Preview: 
https://deploy-preview-38552--osdocs.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.html

https://deploy-preview-38552--osdocs.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/backing_up_and_restoring/restoring-applications.html

Verified by QE. Ready for peer review.

Update: There might be more comments from QE but peer review should still go ahead because this PR is so big.

### CODE REVIEW

File: modules/oadp-backing-up-applications.adoc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// * backup_and_restore/application_backup_and_restore/backing-up-and-restoring-applications.adoc
+
+[id="oadp-backing-up-applications_{context}"]
+= Backing up applications
+
+You back up applications by creating a `Backup` custom resource (CR).
+
+.Prerequisites
+
+* You must have the Data Protection Application installed.
```
Comment (shawn-hurley) [inline modules/oadp-backing-up-applications.adoc:12]: It might be worth noting that you must also have created a DPA CR manifest that is successful?

File: modules/oadp-restoring-applications.adoc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// * backup_and_restore/application_backup_and_restore/backing-up-and-restoring-applications.adoc
+
+[id="oadp-restoring-applications_{context}"]
+= Restoring applications
+
+You restore applications by creating a `Restore` custom resource (CR).
+
+.Prerequisites
+
+* Velero `Backup` CR created.
```
Comment (shawn-hurley) [inline modules/oadp-restoring-applications.adoc:12]: It would be good to add the prerequisites of both the OADP operator and DPA CR here.

Also, if there are backups already in the BSL, then velero will sync backups to this cluster. Maybe just that a velero backup resource must exist?

**Review by shawn-hurley — COMMENTED** [review]
Just some thoughts

File: modules/oadp-backing-up-applications.adoc (hunk 1 above)
Comment (apinnick) [reply modules/oadp-backing-up-applications.adoc:12]: If the DPA CR is not successful, the Data Protect Application would not be considered "installed". I will think about this.

File: backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+You can create xref:../../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc#oadp-creating-backup-hooks_backing-up-applications[backup hooks] to run commands before or after the backup operation.
+
+You can xref:../../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc#oadp-scheduling-backups_backing-up-applications[schedule backups] by creating a `Schedule` CR.
+
+You can back up applications on NFS data volumes with xref:../../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc#oadp-backing-up-applications-restic_backing-up-applications[Restic].
+
+You can back up PVs by using the xref:../../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc#oadp-backing-up-pvs-csi_backing-up-applications[Container Snapshot Interface (CSI)] if the cloud provider supports CSI and does not have a default Velero plugin.
```
Comment (XinRedhat) [inline backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc:16]: >if the cloud provider supports CSI and does not have a default Velero plugin.

Maybe below description is more make sense. 
`if the storage  supports CSI driver and does not have default Velero plugin`, it can include the cloud provider and storage such like OCS 4

File: modules/oadp-creating-backup-cr.adoc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+You back up Kubernetes images, internal images, and persistent volumes (PVs) by creating a `Backup` custom resource (CR).
+
+.Prerequisites
+
+* Backup prerequisites:
+** You must have S3 object storage configured for Velero.
+** You must have a backup location configured in the `DataProtectionApplication` CR.
+* Snapshot prerequisites:
+** You must have a cloud provider that supports snapshots or the Cloud Snapshot Interface (CSI).
```
Comment (XinRedhat) [inline modules/oadp-creating-backup-cr.adoc:16]: >the Cloud Snapshot Interface (CSI).

1. CSI = `Container Snapshot Interface`
2. It seems the description doesn't consider using CSI driver for OCS 4 snapshot

Comment (XinRedhat) [reply modules/oadp-creating-backup-cr.adoc:16]: other looks good to me  @apinnick

Comment (apinnick) [reply modules/oadp-creating-backup-cr.adoc:16]: CSI - good catch. I will check all the other PRs to make sure I didn't do the same typo.

Comment (apinnick) [reply modules/oadp-creating-backup-cr.adoc]: I added a reference to the CSI driver and OCS 4

**Comment by shawn-hurley** [comment]
Screen Shot 2022-01-05 at 3 34 42 PM

I am a little concerned with this warning. The wording is correct, I am just worried that someone might take this as "OADP is tech preview". Do we have a listing of the tech previews features page or maybe a breakdown of the different APIs that are installed and we can add the warning there? 

If others think this is not a real concern then we can move off it :)

**Comment by apinnick** [comment]
> Screen Shot 2022-01-05 at 3 34 42 PM
> 
> I am a little concerned with this warning. The wording is correct, I am just worried that someone might take this as "OADP is tech preview". Do we have a listing of the tech previews features page or maybe a breakdown of the different APIs that are installed and we can add the warning there?
> 
> If others think this is not a real concern then we can move off it :)

Shawn, I see your point. I can reword it to "The cloud storage API for backups" or "The cloud storage API" to remove that implication.

File: backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+:context: backing-up-applications
+
+toc::[]
+
+You back up applications by xref:../../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc#oadp-creating-backup-cr_backing-up-applications[creating a `Backup` custom resource (CR)]. The `Backup` CR creates backup files for Kubernetes resources and internal images and snapshots for persistent volumes (PVs), if the prerequisites are met.
+
+:FeatureName: The cloud storage API for backups
+include::modules/technology-preview.adoc[]
+
+You can back up PVs by using the xref:../../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc#oadp-backing-up-pvs-csi_backing-up-applications[Container Snapshot Interface (CSI)] if the cloud provider supports CSI and does not have a default Velero plugin, for example, OpenShift Container Storage 4.
```
Comment (shawn-hurley) [inline backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc:13]: I believe that it is not just CSI, it is CSI Snapshots which is an optional capability that CSI drivers may or may not support.

Comment (shawn-hurley) [reply backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc:13]: I also wonder if there is a CSI section in the docs that we can link to that would be helpful to users as well?

File: backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc [hunk 6]
```diff
@@ -0,0 +1,25 @@
+[id="backing-up-applications"]
+= Backing up applications
+include::modules/common-attributes.adoc[]
+:context: backing-up-applications
+
+toc::[]
+
+You back up applications by xref:../../../backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc#oadp-creating-backup-cr_backing-up-applications[creating a `Backup` custom resource (CR)]. The `Backup` CR creates backup files for Kubernetes resources and internal images and snapshots for persistent volumes (PVs), if the prerequisites are met.
+
+:FeatureName: The cloud storage API for backups
```
Comment (shawn-hurley) [inline backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc:10]: nit: `CloudStorage` API or `CloudStorage` API for S3 like storage.

**Review by shawn-hurley — COMMENTED** [review]
can we link out to the velero documentation?

File: backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc (hunk 5 above)
Comment (apinnick) [reply backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc:13]: > I also wonder if there is a CSI section in the docs that we can link to that would be helpful to users as well?

CSI section in which docs? OCS? Outside Red Hat docs?

Comment (apinnick) [reply backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc:13]: > I believe that it is not just CSI, it is CSI Snapshots which is an optional capability that CSI drivers may or may not support.

Good point. I didn't realize that the snapshots were optional for CSI drivers. Will fix.

Comment (shawn-hurley) [reply backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc:13]: https://docs.openshift.com/container-platform/4.9/storage/container_storage_interface/persistent-storage-csi-snapshots.html is what I was thinking??

**Comment by apinnick** [comment]
> can we link out to the velero documentation?

Yes, I think we can because 1.7 is a supported version. Where did you want the link?

**Comment by apinnick** [comment]
> I don't know, was just asking. Probably makes more sense in the install section when we say "installs velero 1.7" ?
> 
> Which you already have it there man_facepalming. Probably should read more before commenting sorry :)

I will check. I did include references to Velero but they are in different places.

**Comment by mperetzred** [comment]
@apinnick,
Few comments:
1. here: [link](https://deploy-preview-38552--osdocs.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.html#:~:text=featureFlags:,EnableCSI) featureFlag section is redundant because when adding csi as default plugin is uses that flag behind the scenses.
2. here: [link](https://deploy-preview-38552--osdocs.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.html#:~:text=Optional:%20Add%20the%20defaultVolumesToRestic:%20true%20key-value%20pair%20if%20you%20are%20backing%20up%20volumes%20with%20Restic.), don't see why specifying it here, as the whole template of the schedule should be similar to backup spec, so other backup options should apply there too...

**Comment by apinnick** [comment]
> @apinnick, Few comments:
> 
>     1. here: [link](https://deploy-preview-38552--osdocs.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.html#:~:text=featureFlags:,EnableCSI) featureFlag section is redundant because when adding csi as default plugin is uses that flag behind the scenses.

@XinRedhat Can you confirm? If csi plugin is used, does the user need to set the flag or not?

**Comment by apinnick** [comment]
>     2. here: [link](https://deploy-preview-38552--osdocs.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.html#:~:text=Optional:%20Add%20the%20defaultVolumesToRestic:%20true%20key-value%20pair%20if%20you%20are%20backing%20up%20volumes%20with%20Restic.), don't see why specifying it here, as the whole template of the schedule should be similar to backup spec, so other backup options should apply there too...

@mperetzred  I don't understand what you are saying. What should not be specified?

File: modules/oadp-creating-backup-cr.adoc [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+[id="oadp-creating-backup-cr_{context}"]
+= Creating a Backup CR
+
+You back up Kubernetes images, internal images, and persistent volumes (PVs) by creating a `Backup` custom resource (CR).
+
+.Prerequisites
+
+* You must install the OpenShift API for Data Protection (OADP) Operator.
+* The `DataProtectionApplication` CR must be in a `Ready` state.
```
Comment (mburke5678) [inline modules/oadp-creating-backup-cr.adoc:13]: Do we need to mention that this CR is associated with OADP? Or will users know what this CR is?

Comment (apinnick) [reply modules/oadp-creating-backup-cr.adoc:13]: Yes. The installation section is in a separate PR. It includes creating the DPA CR.

File: modules/oadp-creating-backup-cr.adoc [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+* Backup prerequisites:
+** You must have S3 object storage configured for Velero.
```
Comment (mburke5678) [inline modules/oadp-creating-backup-cr.adoc:15]: Maybe this is my ignorance. Do we need to mention what Velero is and how to determine if everything is configured properly?

File: modules/oadp-creating-backup-cr.adoc [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+** You must have a backup location configured in the `DataProtectionApplication` CR.
```
Comment (mburke5678) [inline modules/oadp-creating-backup-cr.adoc:16]: Is it typical to have a backup location in the CR? Do we need to tell users how to verify?

File: modules/oadp-creating-backup-cr.adoc [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+* Snapshot prerequisites:
+** You must have a cloud provider that uses a native snapshot API or the Container Snapshot Interface (CSI) to create snapshots, such as OpenShift Container Storage 4.
+** You must have a volume location configured in the `DataProtectionApplication` CR.
```
Comment (mburke5678) [inline modules/oadp-creating-backup-cr.adoc:19]: Is it typical to have a volume location in the CR? Do we need to tell users how to verify?

File: modules/oadp-creating-backup-cr.adoc [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+.Prerequisites
+
+* You must install the OpenShift API for Data Protection (OADP) Operator.
+* The `DataProtectionApplication` CR must be in a `Ready` state.
+* Backup prerequisites:
+** You must have S3 object storage configured for Velero.
+** You must have a backup location configured in the `DataProtectionApplication` CR.
+* Snapshot prerequisites:
+** You must have a cloud provider that uses a native snapshot API or the Container Snapshot Interface (CSI) to create snapshots, such as OpenShift Container Storage 4.
```
Comment (mburke5678) [inline modules/oadp-creating-backup-cr.adoc:18]: Not clear to me. 
_You must have a cloud provider that uses a native snapshot API or a cloud provider that uses the Container Snapshot Interface (CSI) to create snapshots...._
 -or-
 _You must have a cloud provider that uses a native snapshot API or you must have the Container Snapshot Interface (CSI) to create snapshots...._
 ??

File: modules/oadp-creating-backup-cr.adoc [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+// * backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc
+
+[id="oadp-creating-backup-cr_{context}"]
+= Creating a Backup CR
+
+You back up Kubernetes images, internal images, and persistent volumes (PVs) by creating a `Backup` custom resource (CR).
+
+.Prerequisites
+
+* You must install the OpenShift API for Data Protection (OADP) Operator.
```
Comment (mburke5678) [inline modules/oadp-creating-backup-cr.adoc:12]: Does the user need the Data Protection Application as required for backing up persistent volumes?

File: modules/oadp-backing-up-pvs-csi.adoc [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+[id="oadp-backing-up-pvs-csi_{context}"]
+= Backing up persistent volumes with Container Snapshot Interface snapshots
+
+You back up persistent volumes with Container Snapshot Interface (CSI) snapshots by creating a `VolumeSnapshotClass` CR.
+
+.Prerequisites
+
+* The cloud provider must support CSI snapshots.
+* You must install the OpenShift API for Data Protection (OADP) Operator and the Data Protection Application.
+* You must add the `csi` default plugin and the `EnableCSI` feature flag to the `DataProtectionApplication` manifest:
```
Comment (mburke5678) [inline modules/oadp-backing-up-pvs-csi.adoc:14]: s/`DataProtectionApplication` manifest/`DataProtectionApplication` CR ??

File: modules/oadp-backing-up-pvs-csi.adoc [hunk 14]
```diff
@@ -0,0 +1,95 @@
+// Module included in the following assemblies:
+//
+// * backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc
+
+[id="oadp-backing-up-pvs-csi_{context}"]
+= Backing up persistent volumes with Container Snapshot Interface snapshots
+
+You back up persistent volumes with Container Snapshot Interface (CSI) snapshots by creating a `VolumeSnapshotClass` CR.
```
Comment (mburke5678) [inline modules/oadp-backing-up-pvs-csi.adoc:8]: s/CR/ custom resource (CR) ? This is the first reference in the module. I think we need to spell it out.

File: modules/oadp-backing-up-pvs-csi.adoc [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+.Prerequisites
+
+* The cloud provider must support CSI snapshots.
+* You must install the OpenShift API for Data Protection (OADP) Operator and the Data Protection Application.
+* You must add the `csi` default plugin and the `EnableCSI` feature flag to the `DataProtectionApplication` manifest:
++
+.Example DataProtectionApplication manifest with CSI
```
Comment (mburke5678) [inline modules/oadp-backing-up-pvs-csi.adoc:16]: If you change manifest to CR above, change here also.

File: modules/oadp-backing-up-pvs-csi.adoc [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+      - openshift
+      - csi
+    featureFlags:
+    - EnableCSI
+----
+
+.Procedure
+
+. Create a `VolumeSnapshotClass` manifest for CSI, as in the following example:
++
```
Comment (mburke5678) [inline modules/oadp-backing-up-pvs-csi.adoc:36]: s/CR/manifest ?

File: modules/oadp-backing-up-pvs-csi.adoc [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+.Example output
++
+[source,terminal]
+----
+NAME              PHASE       LAST VALIDATED   AGE   DEFAULT
+velero-sample-1   Available   11s              31m
+----
+
+. Create a `Backup` CR, as in the following example:
++
```
Comment (mburke5678) [inline modules/oadp-backing-up-pvs-csi.adoc:67]: If you did the previous section, Creating a Backup CR, you already have a `Backup` CR named '<backup>`. I should think that you will not be allowed to do this here. Should the user *edit* the `Backup` CR? 

Also, is there a reason for using the brackets in the name? Seems unusual.

File: modules/oadp-backing-up-pvs-csi.adoc [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+spec:
+  hooks: {}
+  includedNamespaces:
+  - <namespace> <1>
+  includeClusterResources: true
+  storageLocation: <velero-sample-1> <2>
+  ttl: 720h0m0s
+EOF
+----
+<1> Specify an array of namespaces to back up.
```
Comment (mburke5678) [inline modules/oadp-backing-up-pvs-csi.adoc:87]: Will users know how to specify an array? (I'm not sure I do, but I'm not the target audience). 
Should you add more examples to show how to configure? 

  includedNamespaces:
  - <namespace>
  - <namespace2>

File: modules/oadp-backing-up-pvs-csi.adoc (hunk 13 above)
Comment (apinnick) [reply modules/oadp-backing-up-pvs-csi.adoc:14]: A manifest is not the same thing as a CR.

File: modules/oadp-backing-up-pvs-csi.adoc (hunk 17 above)
Comment (apinnick) [reply modules/oadp-backing-up-pvs-csi.adoc:67]: You mean <name>? It's a user-editable value.

File: modules/oadp-backing-up-pvs-csi.adoc (hunk 18 above)
Comment (apinnick) [reply modules/oadp-backing-up-pvs-csi.adoc:87]: Cluster admins know how to create an array. It's just a list. We use them all the time. No need to include example.

File: modules/oadp-backing-up-pvs-csi.adoc (hunk 16 above)
Comment (apinnick) [reply modules/oadp-backing-up-pvs-csi.adoc:36]: The CR is created from the manifest. You create a manifest. That creates the CR.

File: modules/oadp-backing-up-pvs-csi.adoc (hunk 17 above)
Comment (apinnick) [reply modules/oadp-backing-up-pvs-csi.adoc:67]: I could have left out the steps for getting the backupStorageLocations and creating a Backup CR. However, we can't put xrefs in modules and it was only a couple lines.

Maybe there is a way to remove the repetition. I'll try to reorganize it a little.

File: modules/oadp-scheduling-backups.adoc [hunk 19]
```diff
@@ -0,0 +1,64 @@
+// Module included in the following assemblies:
+//
+// * backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc
+
+[id="oadp-scheduling-backups_{context}"]
+= Scheduling backups
+
+You schedule backups by creating a `Schedule` custom resource (CR) instead of a `Backup` CR.
```
Comment (mburke5678) [inline modules/oadp-scheduling-backups.adoc:8]: Should you mention in the assembly that you create the Schedule CR *instead* of the Backup CR?

_You can schedule backups by creating a Schedule CR._

File: modules/oadp-scheduling-backups.adoc [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+.Prerequisites
+
+* You must install the OpenShift API for Data Protection (OADP) Operator.
```
Comment (mburke5678) [inline modules/oadp-scheduling-backups.adoc:12]: Does the user need the Data Protection Application?

**Comment by mburke5678** [comment]
@apinnick This looks good. I had some comments that are likely due to my ignorance. Feel free to disregard anything that would be obvious to a typical user.

File: modules/oadp-backing-up-applications-restic.adoc [hunk 21]
```diff
@@ -0,0 +1,33 @@
+// Module included in the following assemblies:
+//
+// * backup_and_restore/application_backup_and_restore/backing_up_and_restoring/backing-up-applications.adoc
+
+[id="oadp-backing-up-applications-restic_{context}"]
+= Backing up applications with Restic
+
```
Comment (apinnick) [inline modules/oadp-backing-up-applications-restic.adoc:7]: @mburke5678 
I updated this module to remove some of the repetition.

There's no way to verify that the Backup was done with Restic until the backup runs. Or maybe by checking the activity of the Restic pods but that would be overkill.

File: modules/oadp-backing-up-pvs-csi.adoc (hunk 13 above)
Comment (mburke5678) [reply modules/oadp-backing-up-pvs-csi.adoc]: So the the `DataProtectionApplication` CR. and the `DataProtectionApplication` manifest are two separate, distinct objects? Same with `Backup` CR and `Backup` manifest. .

File: modules/oadp-backing-up-applications-restic.adoc (hunk 21 above)
Comment (apinnick) [reply modules/oadp-backing-up-applications-restic.adoc:7]: @mburke5678 Could you please have a look at this module and see whether it makes sense?

**Comment by apinnick** [comment]
Adding label that peer review is done based on mburke in Slack channel.
